heron/stmgr/src/cpp/manager/stateful-restorer.cpp

```diff
@@ -0,0 +1,260 @@
+const sp_string METRIC_INSTANCE_RESTORE_RESPONSES_IGNORED = "__instance_restore_response_ignored";
+
+StatefulRestorer::StatefulRestorer(CkptMgrClient* _ckptmgr,
+                             StMgrClientMgr* _clientmgr, TupleCache* _tuple_cache,
+                             StMgrServer* _server,
+                             common::MetricsMgrSt* _metrics_manager_client,
+                             std::function<void(proto::system::StatusCode,
+                                                std::string, sp_int64)> _restore_done_watcher) {
+  ckptmgr_ = _ckptmgr;
+  clientmgr_ = _clientmgr;
+  tuple_cache_ = _tuple_cache;
+  server_ = _server;
+
+  in_progress_ = false;
+  restore_done_watcher_ = _restore_done_watcher;
+  metrics_manager_client_ = _metrics_manager_client;
+  multi_count_metric_ = new common::MultiCountMetric();
+  time_spent_metric_ = new common::TimeSpentMetric();
+  metrics_manager_client_->register_metric("__stateful_restore_count", multi_count_metric_);
+  metrics_manager_client_->register_metric("__stateful_restore_time", time_spent_metric_);
+}
+
+StatefulRestorer::~StatefulRestorer() {
+  metrics_manager_client_->unregister_metric("__stateful_restore_count");
+  metrics_manager_client_->unregister_metric("__stateful_restore_time");
+  delete multi_count_metric_;
+  delete time_spent_metric_;
+}
+
+void StatefulRestorer::StartRestore(const std::string& _checkpoint_id, sp_int64 _restore_txid,
+                                    proto::system::PhysicalPlan* _pplan) {
+  multi_count_metric_->scope(METRIC_START_RESTORE)->incr();
+  if (in_progress_) {
+    multi_count_metric_->scope(METRIC_START_RESTORE_IN_PROGRESS)->incr();
+    LOG(WARNING) << "Got a RestoreTopologyState request for " << _checkpoint_id
+                 << " " << _restore_txid << " while we were still in old one "
+                 << checkpoint_id_ << " " << restore_txid_;
+    if (_restore_txid <= restore_txid_) {
+      LOG(FATAL) << "New restore txid: " << _restore_txid << " is <= old "
+                 << restore_txid_<< "!!! Dying!!! ";
+    }
+  } else {
+    LOG(INFO) << "Starting Restore for checkpoint_id " << _checkpoint_id
+              << " and txid " << _restore_txid;
+    tuple_cache_->clear();
+    clientmgr_->CloseConnectionsAndClear();
+    server_->ClearCache();
+    time_spent_metric_->Start();
+  }
+  // This is a new one for this checkpoint
+  in_progress_ = true;
+  clients_connections_pending_ = true;
+  instance_connections_pending_ = true;
+  std::vector<proto::system::Instance*> instances;
+  server_->GetInstanceInfo(instances);
+  local_taskids_.clear();
+  for (auto instance : instances) {
+    local_taskids_.insert(instance->info().task_id());
+  }
+  restore_pending_ = local_taskids_;
+  get_ckpt_pending_ = local_taskids_;
+  checkpoint_id_ = _checkpoint_id;
+  restore_txid_ = _restore_txid;
+
+  // Send messages to ckpt
+  GetCheckpoints();
```

Comment:
why's the comment discuss sending message but the method implies getting a checkpoint?